### PR TITLE
build: create JS bundles for each component

### LINKED ISSUE #4137: Tree-shaking for production build is not working properly
#### Bug, feature request, or proposal:
The production bundle size is not shrinking
 
#### What is the expected behavior?
Base on the new package structure the production bundle (ng build --prod) should not have included all the Material Components Module

#### What is the current behavior?
Base on the change log now we should create our own Custom MaterialModule, in my case I have this one:
```ts
import { NgModule } from '@angular/core';
import {
    MdCardModule,
    MdChipsModule,
    MdToolbarModule,
    MdButtonModule,
    MdSidenavModule,
    MdRippleModule,
    MdIconModule,
    MdListModule,
    MdProgressSpinnerModule,
} from '@angular/material';

@NgModule({
    imports: [
        MdCardModule,
        MdChipsModule,
        MdToolbarModule,
        MdButtonModule,
        MdSidenavModule,
        MdRippleModule,
        MdIconModule,
        MdListModule,
        MdProgressSpinnerModule,
    ],
    declarations: [

    ],
    providers: [

    ],
    exports: [
        MdCardModule,
        MdChipsModule,
        MdToolbarModule,
        MdButtonModule,
        MdSidenavModule,
        MdRippleModule,
        MdIconModule,
        MdListModule,
        MdProgressSpinnerModule,
    ]
})
export class CustomMaterialModule { }
```
that is imported in the AppModule.

Even if I remove all the module and import only, for example, MdButtonModule, the size of material library in the production vendor.js still 308Kb 

#### Which versions of Angular, Material, OS, browsers are affected?
@angular/cli: 1.0.0
node: 7.9.0
os: linux x64
@angular/animations: 4.0.1
@angular/common: 4.0.1
@angular/compiler: 4.0.1
@angular/core: 4.0.1
@angular/forms: 4.0.1
@angular/http: 4.0.1
@angular/material: 2.0.0-beta.3
@angular/platform-browser: 4.0.1
@angular/platform-browser-dynamic: 4.0.1
@angular/router: 4.0.1
@angular/cli: 1.0.0
@angular/compiler-cli: 4.0.1

#### Is there anything else we should know?
This is the vendor sourcemap generated by source-map-explorer
image

### PR DESCRIPTION
**Pull request by jelbourn**
Before this change, all of `@angular/material` was bundled into a
single, flat JS file.

After this change, each top-level component (i.e., each directory under
`src/lib`) will be have its own JS file (and d.ts and metadata.json).

The benefit of this change is that current versions of webpack will be
able to do a much better job of only including the code an app is using.

It also allows (but does not require) people to explicitly import from specific components,
e.g.
```ts
import {MdButtonModule} from '@angular/material/button';
```

For development _within Angular Material itself_ we have to use these
deeper imports because the root `@angular/material` doesn't exist until
the actual package is created.

Additionally, this PR adds a `tsconfig.json` for each dir under `src/`
with the correct path mappings for that directory, so that IDEs will
correctly generate imports.

Closure Compiler CI check is also disabled by this PR; it cannot work
with the new package structure until a release with
google/closure-compiler#2600 is made available.

Fixes #4137

### CODE REVIEW

File: src/lib/tsconfig-build.json [hunk 1]
```diff
@@ -5,7 +5,7 @@
     "declaration": true,
     "stripInternal": false,
     "experimentalDecorators": true,
-    "noUnusedParameters": true,
+    "noUnusedParameters": false,
```
Comment (DevVersion) [inline src/lib/tsconfig-build.json]: Why are we no longer having this check? Maybe add a comment if necessary

File: tools/package-tools/entry-point-package-json.ts [hunk 2]
```diff
@@ -8,8 +8,8 @@ export function createEntryPointPackageJson(destDir: string, packageName: string
     name: `@angular/${packageName}/${entryPointName}`,
     typings: `../${entryPointName}.d.ts`,
     main: `../bundles/${packageName}-${entryPointName}.umd.js`,
-    module: `../@angular/${packageName}/${entryPointName}.es5.js`,
-    es2015: `../@angular/${packageName}/${entryPointName}.js`,
+    module: `../esm5/${packageName}/${entryPointName}.es5.js`,
+    es2015: `../esm2015/${packageName}/${entryPointName}.js`,
```
Comment (DevVersion) [inline tools/package-tools/entry-point-package-json.ts]: Out of curiosity: How does this work with Closure Compiler now? Is there an alternative to the `js_module_roots` option?

File: src/lib/tsconfig-umd.json [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    "typings.d.ts"
+  ],
+  "angularCompilerOptions": {
+    "annotateForClosureCompiler": true,
+    "strictMetadataEmit": true,
+    "flatModuleOutFile": "index.js",
+    "flatModuleId": "@angular/material",
+    "skipTemplateCodegen": true
+  }
+}
```
Comment (DevVersion) [inline src/lib/tsconfig-umd.json]: Is that file used anywhere? We now have three `tsconfig-XX` files in the `lib/` package.

Comment (jelbourn) [reply src/lib/tsconfig-umd.json]: I was just trying something out with this, meant to delete it.

File: tools/package-tools/entry-point-package-json.ts (hunk 2 above)
Comment (jelbourn) [reply tools/package-tools/entry-point-package-json.ts]: Short answer: it doesn't.

I updated the script and left a comment explaining the current situation w/ closure.

File: src/lib/tsconfig-build.json (hunk 1 above)
Comment (jelbourn) [reply src/lib/tsconfig-build.json]: Was just trying something during development; reverted

**Comment by jelbourn** [comment]
* Rebased such that it includes `material-moment-adapter/`
* Added a `tsconfig.json` per `src/` directory in order to make IDEs auto-add imports to the right path.
* Disabled closure compiler CI check temporarily (need a new release)
* Refactored rollup bundling code into a class, `PackageBundler` so that it can maintain the context of the `BuildPackage`
* Fix prerender

**Comment by jelbourn** [comment]
@DevVersion after this PR we're definitely going to have to parallelize the typescript builds and bundling. With this change it takes over a minute. To paralleize the tasks, we'll need to manually spawn a child process for each, since nodejs is single-threaded. We also probably want to tweak the watch tasks to only re-compile the sub-packages that changed. Let me know if you want to work on all that.



All of this will be obsolete when we switch to bazel, but we're at least a month away from making that switch.

**Comment by jelbourn** [comment]
@DevVersion the tsconfig in the root would cause IDEs to autogenerate the wrong imports sometimes.

**Comment by jelbourn** [comment]
Caretaker note: both the copy config and the build rules need to updated alongside this change

**Comment by jelbourn** [comment]
Another note: if this gets merged alongside another PR, it will potentially need a build fix (for anything importing from, e.g., `../core`)

**Pull request merged**